validator-network/src/network_impl.rs

```diff
@@ -76,50 +88,111 @@ where
+        self.validator_peer_id_cache
+            .write()
+            .insert(public_key.compressed().clone(), CacheState::InProgress);
+        let cache_value = match Self::resolve_peer_id(&self.network, public_key).await {
+            Ok(result) => match result {
+                Some(peer_id) => {
+                    log::trace!(
+                        %peer_id,
+                        validator_id,
+                        %public_key,
+                        "Resolved validator peer ID"
+                    );
+                    CacheState::Resolved(peer_id)
+                }
+                None => {
+                    log::trace!(validator_id, %public_key, "Unable to resolve validator peer ID: Entry not found in DHT");
+                    CacheState::Error(NetworkError::Unreachable)
+                }
+            },
+            Err(e) => {
+                log::trace!(
+                    validator_id,
+                    error = ?e,
+                    %public_key,
+                    "Unable to resolve validator peer ID: Network error"
+                );
+                CacheState::Error(e)
+            }
+        };
+        self.validator_peer_id_cache
+            .write()
+            .insert(public_key.compressed().clone(), cache_value);
+    }
+
     /// Look up the peer ID for a validator ID.
     async fn get_validator_peer_id(
         &self,
         validator_id: u16,
-    ) -> Result<N::PeerId, NetworkError<N::Error>> {
-        let (peer_id, public_key) = {
-            let public_key = self
-                .validator_keys
-                .read()
-                .get(usize::from(validator_id))
-                .ok_or(NetworkError::UnknownValidator(validator_id))?
-                .clone();
+    ) -> Result<N::PeerId, NetworkError<N::Error>>
+    where
+        <N as Network>::Error: Sync + Send,
+    {
+        let public_key = self
+            .validator_keys
+            .read()
+            .get(usize::from(validator_id))
+            .ok_or(NetworkError::UnknownValidator(validator_id))?
+            .clone();
 
-            let peer_id = self
+        let update_peer_cache = {
+            if let Some(cache_state) = self
                 .validator_peer_id_cache
                 .read()
                 .get(&public_key.compressed().clone())
-                .cloned();
-            (peer_id, public_key)
+            {
+                match cache_state {
+                    CacheState::Resolved(peer_id) => return Ok(*peer_id),
+                    CacheState::Error(_) => true,
+                    CacheState::InProgress => {
+                        log::debug!(validator_id, "Record resolution is in progress");
+                        false
+                    }
+                }
+            } else {
+                log::error!(
+                    ?public_key,
+                    validator_id,
+                    "Could not find peer ID for validator in DHT",
+                );
```

Comment:
Just out of curiosity, feel free to ignore this.
`error!` might be a bit high to me as this seemingly would be hit during normal execution. For example whenever a node starts, or when a new validator gets elected. Since it would be expected behaviour then maybe debug is more appropriate.